### PR TITLE
fix: normalize HTTP MCP workflow inputs

### PR DESCRIPTION
**Pull request by cnYui**
## Summary

- Add a shared MCP input normalizer that parses JSON string roots and recursively restores dense numeric-index records back to arrays.
- Normalize workflow node fields that HTTP MCP clients can serialize incorrectly, including `typeVersion`, `position`, `parameters`, and `credentials`.
- Apply the normalizer to workflow create/update schemas, validation schemas, and partial workflow diff `addNode`/`position` inputs.
- Add regression coverage for the nested serialization failures reported in #814.

## Scope

This addresses the nested serialization portion of #814. It intentionally does not change the separate `PUT /api/v1/workflows/:id` additional-properties behavior, because that appears to be a distinct n8n API compatibility issue.

## Verification

- `npx vitest run tests/unit/services/n8n-validation.test.ts --coverage.enabled=false`
- `npx vitest run tests/unit/mcp/handlers-n8n-manager.test.ts --coverage.enabled=false`
- `npx vitest run tests/unit/mcp/handlers-workflow-diff.test.ts --coverage.enabled=false`
- `npm run typecheck`
- `npm run build`

Note: `npm ci` could not be used locally because `package.json` and `package-lock.json` are currently out of sync. Dependencies were installed with `npm install --package-lock=false --no-audit --no-fund` without changing tracked files.

### CODE REVIEW

**Comment by czlonkowski** [comment]
Thanks @cnYui — this is a well-aimed fix. The normalizer design (dense-index heuristic with the empty-object/sparse-key guards, root-only JSON parsing) is exactly the right deterministic chokepoint for #814, and the regression tests mirror the reported shapes faithfully.

I've pushed two things to your branch (maintainer edit):

1. **Rebase onto current main** (v2.57.2) — no conflicts; this also brings the committed `dist/` build current.
2. **A follow-up commit extending coverage** to the spots the same mangling could still reach:
   - `workflowDiffSchema.operations` root (a dense-record-mangled operations *array* failed before the per-field preprocessors ever ran) plus `updates`, `patches`, `connections` (replaceConnections), and `settings`. The `updateNode.updates` gap was the most important one — mangled arrays there were written verbatim into node parameters via dot-notation, corrupting the workflow silently instead of failing loudly.
   - `listWorkflowsSchema.tags` (`{"0":"tag"}` → `["tag"]`).
   - `normalizeMcpWorkflowNode` no longer introduces explicit `undefined`-valued keys for fields absent from the input.
   - A file-level comment documenting the deliberate heuristic tradeoff, direct unit tests for the normalizer edge cases (sparse keys, nested strings never JSON-parsed, idempotency), and handler regression tests for each newly covered field.

Verification: typecheck, build, and the full test suite pass (the only failures in a full parallel run are known-flaky live-instance integration tests that pass in isolation, unrelated to this change).

Will merge once CI is green. This should also resolve #600, #611, and #492.

File: src/utils/mcp-input-normalizer.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+  return {
+    ...parsed,
+    typeVersion: normalizeNumberLike(parsed.typeVersion),
+    position: normalizeMcpJsonValue(parsed.position),
+    parameters: normalizeMcpJsonValue(parsed.parameters),
+    credentials: parsed.credentials === undefined
+      ? undefined
+      : normalizeMcpJsonValue(parsed.credentials),
+  };
```
[the remark this answers is not in the document]
Comment (czlonkowski) [reply src/utils/mcp-input-normalizer.ts]: Addressed in 228103b — `normalizeMcpWorkflowNode` now only rewrites fields present on the input (no explicit `undefined`-valued keys added for `typeVersion`/`position`/`parameters`/`credentials`), with a unit test asserting no keys are added.

File: src/utils/mcp-input-normalizer.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  if (keys.length === 0) {
+    return false;
+  }
+
+  return keys.every((key) => /^\d+$/.test(key))
+    && keys
+      .map(Number)
+      .sort((a, b) => a - b)
+      .every((key, index) => key === index);
+}
```
[the remark this answers is not in the document]
Comment (czlonkowski) [reply src/utils/mcp-input-normalizer.ts:55]: Good catch — confirmed real: `{"00": "a"}` was converted to `[undefined]`, silently dropping the value. Fixed in 70f1c67: the dense-index check now only accepts canonical array-index keys (`/^(0|[1-9]\d*)$/`), so records with leading-zero keys are left untouched. Regression tests added.

**Comment by czlonkowski** [comment]
CI note: fork PRs don't receive repository secrets, so the integration tests and Docker builds cannot pass here (empty `N8N_API_URL`/registry credentials — same failure as on the original push). I ran the full secrets-enabled suite against this exact head commit via the temporary draft PR #841: **all checks green**, including the 9-minute test job with live n8n integration tests, both Docker builds, CodeQL, and codecov. Run: https://github.com/czlonkowski/n8n-mcp/actions/runs/27264650748

**Comment by czlonkowski** [comment]
Second verification run for the review-hardening head (25398f8) — covering the position de-stringification, canonical number gating, recursion depth cap, and credentials exemption — is **all green** via draft PR #842: full test job (8m12s, live n8n integration included), Docker builds, CodeQL, codecov. The fix was also exercised end-to-end against a live n8n instance through the MCP tools (mangled create/addNode/updateNode payloads, operations-as-dense-record, mangled tags, leading-zero guard, and a well-formed control). Run: https://github.com/czlonkowski/n8n-mcp/actions/runs/27270119239
